packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts

```diff
@@ -0,0 +1,252 @@
+    this.stackResources = props.listStackResources;
+    this.context = {
+      'AWS::AccountId': props.account,
+      'AWS::Region': props.region,
+      'AWS::Partition': props.partition,
+      'AWS::URLSuffix': props.urlSuffix,
+      ...props.parameters,
+    };
+    this.account = props.account;
+    this.region = props.region;
+    this.partition = props.partition;
+  }
+
+  public async findPhysicalNameFor(logicalId: string): Promise<string | undefined> {
+    const stackResources = await this.stackResources.listStackResources();
+    return stackResources.find(sr => sr.LogicalResourceId === logicalId)?.PhysicalResourceId;
+  }
+
+  public async evaluateCfnExpression(cfnExpression: any): Promise<any> {
+    const self = this;
+    class CfnIntrinsics {
+      public evaluateIntrinsic(intrinsic: Intrinsic): any {
+        const intrinsicFunc = (this as any)[intrinsic.name];
+        if (!intrinsicFunc) {
+          throw new CfnEvaluationException(`CloudFormation function ${intrinsic.name} is not supported`);
+        }
+
+        const argsAsArray = Array.isArray(intrinsic.args) ? intrinsic.args : [intrinsic.args];
+
+        return intrinsicFunc.apply(this, argsAsArray);
+      }
+
+      async 'Fn::Join'(separator: string, args: any[]): Promise<string> {
+        const evaluatedArgs = await self.evaluateCfnExpression(args);
+        return evaluatedArgs.join(separator);
+      }
+
+      async 'Fn::Split'(separator: string, args: any): Promise<string> {
+        const evaluatedArgs = await self.evaluateCfnExpression(args);
+        return evaluatedArgs.split(separator);
+      }
+
+      async 'Fn::Select'(index: number, args: any[]): Promise<string> {
+        const evaluatedArgs = await self.evaluateCfnExpression(args);
+        return evaluatedArgs[index];
+      }
+
+      async 'Ref'(logicalId: string): Promise<string> {
+        const refTarget = await self.findRefTarget(logicalId);
+        if (refTarget) {
+          return refTarget;
+        } else {
+          throw new CfnEvaluationException(`Parameter or resource '${logicalId}' could not be found for evaluation`);
+        }
+      }
+
+      async 'Fn::GetAtt'(logicalId: string, attributeName: string): Promise<string> {
+        // ToDo handle the 'logicalId.attributeName' form of Fn::GetAtt
+        const attrValue = await self.findGetAttTarget(logicalId, attributeName);
+        if (attrValue) {
+          return attrValue;
+        } else {
+          throw new CfnEvaluationException(`Attribute '${attributeName}' of resource '${logicalId}' could not be found for evaluation`);
+        }
+      }
+
+      async 'Fn::Sub'(template: string, explicitPlaceholders?: { [variable: string]: string }): Promise<string> {
+        const placeholders = explicitPlaceholders
+          ? await self.evaluateCfnExpression(explicitPlaceholders)
+          : {};
+
+        // this code is a little convoluted, because we need to deal with async operations
```

Comment:
Wouldn't it be nicer to define a helper:

```ts
async function asyncRegexReplace(s: string, regex: RegExp, cb: (x: string) => Promise<string>) {
  if (!regex.global) { throw new Error('Regex must be created with /g flag'); }
  const ret = new Array<string>();

  let start = 0;  
  while (true) {
    const m = regex.exec(s);
    if (!m) { break; }

    ret.push(s.substring(start, m.index));
    ret.push(await cb(m[1]));
    start = regex.lastIndex;
  }

  ret.push(s.substr(start));
  return ret.join('');
 
}

Reply:
Oh wow. Thats kind of nuts. I had no idea about this mutability of regular expressions in JavaScript. That's actually kind of scary 😜.

Changed!